### PR TITLE
Update address validator implementation

### PR DESCRIPTION
**Pull request by timEulitz**

### CODE REVIEW

File: tests/Unit/UseCases/UpdateDonor/UpdateDonorValidatorTest.php [hunk 1]
```diff
@@ -36,19 +35,47 @@ public function testGivenAnonymousDonor_validationFails() {
...
+		$this->assertEquals( 'salutation', $violations[0]->getSource() );
+		$this->assertEquals( 'firstName', $violations[1]->getSource() );
+		$this->assertEquals( 'lastName', $violations[2]->getSource() );
+		$this->assertEquals( 'street', $violations[3]->getSource() );
+		$this->assertEquals( 'postcode', $violations[4]->getSource() );
+		$this->assertEquals( 'city', $violations[5]->getSource() );
+		$this->assertEquals( 'country', $violations[6]->getSource() );
+	}
+
+	private function newUpdateDonorRequest(): UpdateDonorRequest {
```
Comment (gbirke) [inline tests/Unit/UseCases/UpdateDonor/UpdateDonorValidatorTest.php]: maybe call it `newEmptyUpdateDonorRequest `

File: tests/Unit/UseCases/UpdateDonor/UpdateDonorValidatorTest.php [hunk 2]
```diff
@@ -36,19 +35,47 @@ public function testGivenAnonymousDonor_validationFails() {
...
 	public function testGivenFailingDonorValidator_validationFails() {
-		$addressViolation = new ConstraintViolation( '', 'donor_name_missing', 'first_name' );
-		$donorValidator = $this->createPartialMock( DonorAddressValidator::class, ['getViolations', 'validate' ] );
-		$donorValidator->method( 'getViolations' )->willReturn( [ $addressViolation ] );
+		$addressViolation = new ValidationResult( new ConstraintViolation( '', 'donor_name_missing', 'first_name' ) );
+		$donorValidator = $this->createPartialMock(
+			AddressValidator::class,
+			[ 'validatePostalAddress', 'validatePersonName' ]
+		);
+		$donorValidator->method( 'validatePersonName' )->willReturn( $addressViolation );
```
Comment (gbirke) [inline tests/Unit/UseCases/UpdateDonor/UpdateDonorValidatorTest.php]: To be really thorough, you'd need to add 2 more tests:

* an AddressValidator where  `validatePostalAddress ` fails
* an AddressValidator where  `validateCompanyName ` fails + giving it a company name.

File: src/UseCases/UpdateDonor/UpdateDonorValidator.php [hunk 3]
```diff
@@ -15,26 +17,68 @@ class UpdateDonorValidator {
...
-		$this->donorValidator->validate( $donorRequest );
-		if ( $this->donorValidator->getViolations() ) {
-			return new UpdateDonorValidationResult( ...$this->donorValidator->getViolations() );
+
+		$violations = array_merge(
+			$nameViolations,
+			$this->getAddressViolations( $donorRequest ),
+			$this->getEmailViolations( $donorRequest )
+		);
+		if ( $violations ) {
```
Comment (tzhelyazkova) [inline src/UseCases/UpdateDonor/UpdateDonorValidator.php]: `if ( empty( $violations ) )` ?
p.s. I am highly influenced by TypeScript now :no_good_woman:

File: tests/Unit/UseCases/UpdateDonor/UpdateDonorValidatorTest.php [hunk 4]
```diff
@@ -36,19 +38,79 @@ public function testGivenAnonymousDonor_validationFails() {
...
+	private function newInvalidUpdateCompanyDonorRequest(): UpdateDonorRequest {
+		return ( new UpdateDonorRequest() )
+			->withType( DonorName::PERSON_COMPANY )
+			->withCompanyName( str_repeat( 'TEST', 26 ) )
+			->withStreetAddress( str_repeat( 'TEST', 26 ) )
+			->withPostalCode( str_repeat( '1', 10 ) )
+			->withCity( str_repeat( 'TEST', 26 ) )
+			->withCountryCode( str_repeat( 'TEST', 26 ) )
+			->withEmailAddress( '' );
+	}
```
Comment (tzhelyazkova) [inline tests/Unit/UseCases/UpdateDonor/UpdateDonorValidatorTest.php]: what an aggressive set up :laughing:

File: tests/Integration/UseCases/AddDonation/AddDonationValidatorTest.php [hunk 5]
```diff
@@ -111,7 +112,8 @@ public function testGivenFailingBankDataValidator_validatorReturnsFalse(): void
 			new PaymentDataValidator( 1.0, 100000, [ PaymentMethod::DIRECT_DEBIT ] ),
 			$bankDataValidator,
 			$this->newEmptyIbanBlocklist(),
-			$this->newEmailValidator()
+			$this->newEmailValidator(),
+			$this->newAddressValidator()
```
Comment (tzhelyazkova) [inline tests/Integration/UseCases/AddDonation/AddDonationValidatorTest.php]: Genuine question: There was no address validation in this bank data related test before either way, so why add it now? 
(same goes for the few test below )

File: tests/Unit/UseCases/UpdateDonor/UpdateDonorValidatorTest.php (hunk 4 above)
Comment (timEulitz) [reply tests/Unit/UseCases/UpdateDonor/UpdateDonorValidatorTest.php]: WHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHATWHAT

File: tests/Integration/UseCases/AddDonation/AddDonationValidatorTest.php (hunk 5 above)
Comment (timEulitz) [reply tests/Integration/UseCases/AddDonation/AddDonationValidatorTest.php]: It was already there but the previous solution was instantiated in the constructor:

https://github.com/wmde/fundraising-donations/blob/d926faf6fb545531fb433cd106ec3a252aabb5f3/src/UseCases/AddDonation/AddDonationValidator.php#L54

**Comment by timEulitz** [comment]
@gbirke Merging now but if you see issues later, please let me know.

**Pull request merged**